sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/LowLatencyMemoryStream.scala

```diff
@@ -0,0 +1,300 @@
+  SupportsRealTimeRead
+}
+import org.apache.spark.sql.connector.read.streaming.SupportsRealTimeRead.RecordStatus
+import org.apache.spark.sql.execution.datasources.v2.LowLatencyClock
+import org.apache.spark.sql.execution.streaming.runtime._
+import org.apache.spark.util.{Clock, RpcUtils}
+
+/**
+ * A low latency memory source from memory, only for unit test purpose.
+ * This class is very similar to ContinuousMemoryStream, except that it implements the
+ * interface of SupportsRealTimeMode, rather than ContinuousStream
+ * The overall strategy here is:
+ *  * LowLatencyMemoryStream maintains a list of records for each partition. addData() will
+ *    distribute records evenly-ish across partitions.
+ *  * RecordEndpoint is set up as an endpoint for executor-side
+ *    LowLatencyMemoryStreamInputPartitionReader instances to poll. It returns the record at
+ *    the specified offset within the list, or null if that offset doesn't yet have a record.
+ */
+class LowLatencyMemoryStream[A: Encoder](
+    id: Int,
+    sqlContext: SQLContext,
+    numPartitions: Int = 2,
+    clock: Clock = LowLatencyClock.getClock)
+    extends MemoryStreamBaseClass[A](0, sqlContext)
+    with SupportsRealTimeMode {
+  private implicit val formats: Formats = Serialization.formats(NoTypeHints)
+
+  // LowLatencyReader implementation
+
+  @GuardedBy("this")
+  private val records = Seq.fill(numPartitions)(new ListBuffer[UnsafeRow])
+
+  private val recordEndpoint = new ContinuousRecordEndpoint(records, this)
+  @volatile private var endpointRef: RpcEndpointRef = _
+
+  override def addData(data: IterableOnce[A]): Offset = synchronized {
+    // Distribute data evenly among partition lists.
+    data.iterator.to(Seq).zipWithIndex.map {
+      case (item, index) =>
+        records(index % numPartitions) += toRow(item).copy().asInstanceOf[UnsafeRow]
+    }
+
+    // The new target offset is the offset where all records in all partitions have been processed.
+    LowLatencyMemoryStreamOffset((0 until numPartitions).map(i => (i, records(i).size)).toMap)
+  }
+
+  def addData(partitionId: Int, data: IterableOnce[A]): Offset = synchronized {
+    require(
+      partitionId >= 0 && partitionId < numPartitions,
+      s"Partition ID $partitionId is out of bounds for $numPartitions partitions."
+    )
+
+    // Add data to the specified partition.
+    records(partitionId) ++= data.iterator.map(item => toRow(item).copy().asInstanceOf[UnsafeRow])
+
+    // The new target offset is the offset where all records in all partitions have been processed.
+    LowLatencyMemoryStreamOffset((0 until numPartitions).map(i => (i, records(i).size)).toMap)
+  }
+
+  override def initialOffset(): OffsetV2 = {
+    LowLatencyMemoryStreamOffset((0 until numPartitions).map(i => (i, 0)).toMap)
+  }
+
+  override def latestOffset(startOffset: OffsetV2, limit: ReadLimit): OffsetV2 = {
+    LowLatencyMemoryStreamOffset((0 until numPartitions).map(i => (i, records(i).size)).toMap)
```

Comment:
Do we also need to add `synchronized` to `latestOffset` as it also access `records`? Although it just reads, but the `size` might be inconsistent.